internal/service/synthetics/retry.go

```diff
@@ -0,0 +1,74 @@
+package synthetics
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/synthetics"
+	"github.com/hashicorp/aws-sdk-go-base/v2/awsv1shim/v2/tfawserr"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
+)
+
+const (
+	canaryCreateFail = "CREATE_FAILED"
+)
+
+func retryCreateCanary(conn *synthetics.Synthetics, d *schema.ResourceData, input *synthetics.CreateCanaryInput) (*synthetics.Canary, error) {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{synthetics.CanaryStateCreating, synthetics.CanaryStateUpdating},
+		Target:  []string{synthetics.CanaryStateReady},
+		Refresh: statusCanaryState(conn, d.Id()),
+		Timeout: canaryCreatedTimeout,
+	}
+
+	outputRaw, err := stateConf.WaitForState()
+	if output, ok := outputRaw.(*synthetics.Canary); ok {
+		if status := output.Status; aws.StringValue(status.State) == synthetics.CanaryStateError && aws.StringValue(status.StateReasonCode) == canaryCreateFail {
+			// delete canary because it is the only way to reprovision if in an error state
+			err = deleteCanary(conn, d.Id())
+			if err != nil {
```

Comment:
Should we just `return err` if `err != nil`? If the canary couldn't be deleted is it worth trying to create it again?
Same for the other error checks in this block.